stl/inc/xutility

```diff
@@ -5767,6 +5768,121 @@ private:
     _Udiff _Bmask; // 2^_Bits - 1
 };
 
+template <class _Diff, class _Urng>
+class _Rng_from_urng_v2 { // wrap a URNG as an RNG
+public:
+    using _Ty0 = make_unsigned_t<_Diff>;
+    using _Ty1 = typename _Urng::result_type;
+
+    using _Udiff                              = conditional_t<sizeof(_Ty1) < sizeof(_Ty0), _Ty0, _Ty1>;
+    constexpr static unsigned int _Udiff_bits = sizeof(_Udiff) * CHAR_BIT;
+    using _Uprod = conditional_t<_Udiff_bits <= 16, uint32_t, conditional_t<_Udiff_bits <= 32, uint64_t, _Unsigned128>>;
+
+    explicit _Rng_from_urng_v2(_Urng& _Func) : _Ref(_Func) {}
+
+    _Diff operator()(_Diff _Index) { // adapt _Urng closed range to [0, _Index)
+        // From Daniel Lemire, "Fast Random Integer Generation in an Interval", ACM Trans. Model. Comput. Simul. 29 (1),
+        // 2019.
+        _Udiff _Mask;
+        unsigned int _Niter;
+
+        if constexpr (_Bits < _Udiff_bits) {
+            _Mask  = 0;
+            _Niter = 0;
+            do {
+                _Mask <<= _Bits;
+                _Mask |= _Bmask;
+                ++_Niter;
+            } while (_Mask < _Udiff(_Index - 1));
+        } else {
+            _Mask  = _Bmask;
+            _Niter = 1;
+        }

```

Comment:
> Unless I am confused, in the if constexpr case, we will always end up with an _Niter of 2 or more.

No, the loop condition is `while (_Mask < static_cast<_Udiff>(_Index - 1))`, so the number of iterations depends on the input parameter. The `else` paths are all for the situation where the URBG creates at least as many bits as are in `_Udiff`, so that we know we only need one iteration, regardless or the input.

I can do a `while`, it just tends to produce slightly larger code, because the compiler duplicates the `while` condition, once at the head of the loop to decide whether to enter, and again at the tail to decide whether to jump back. So I was hoping that less jumpy code would be faster, but as I recall it didn't make much difference. I'll make the change for now and rerun my benchmark when I'm done with the rest of the feedback.